text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+- Feature Name: `direct_enum_discriminant`
+- Start Date: 2024-03-16
+- RFC PR: [rust-lang/rfcs#3607](https://github.com/rust-lang/rfcs/pull/3607)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Enable using **`.enum#discriminant`** on values of enum type from safe code in the same module
+to get the numeric value of the variant's discriminant in the numeric type of its `repr`.
+
+# Motivation
+[motivation]: #motivation
+
+Today in Rust you can use `as` casts on *field-less* `enum`s to get their discriminants,
+but as soon as any variant has fields, that's no longer available.
```

Comment:
Worse: As soon as an enum has any tuple-like variants, `as` casting the tuple-like variant will still work, but instead of producing the discriminant, it will instead cast the function pointer of that variant's constructor. This, I think, is the real reason why we shouldn't "just" expand the functionality of `as` casting to fieldful enums; it further entrenches an exceptionally dangerous pattern.